hash/hash_comparator.go

```diff
@@ -260,3 +290,14 @@ func decodeScryptHash(encodedHash string) (p *Scrypt, salt, hash []byte, err err
 
 	return p, salt, hash, nil
 }
+
+// decodeMD5Hash decodes MD5 encoded password hash.
+// format: $md5$<hash>
+func decodeMD5Hash(encodedHash string) (hash []byte, err error) {
+	parts := strings.Split(encodedHash, "$")
```

Comment:
Most md5 hashers probably use salting, but there's probably not one way to add the salt (as MD5 itself has no salting). How could we solve this?

Reply:
Yeah that would be great addition. In most of the scenarios i have seen, salt is either appended or prepended to the password. And again probably not everyone uses salting. I think format like below, can accommodate up to that needs:
`$md5$<hash>$<salt base64>,<mode>`
- **mode** parameter can be either **a** *(abbreviation for append)* or **p** *(abbreviation for prepend)*. In the case of **a** salt will be appended to the password before it's being hashed. In the case of **p** salt will be prepended to the password before it's being hashed.
- The section after the third dollar sign (inclusive) is optional. That means if someone wants to store a password which has no salt, it can be formatted like this: `$md5$<hash>`. By putting **hash** before **salt** we can easily make *salting* optional.

Reply:
In that case,
Password without a salt can look like this: `$md5$v=1$m=n$$202cb962ac59075b964b07152d234b70`
Password with salt can look like this: `$md5$v=1$m=a$MTIzNDU2Nw==$8a3379ae294935afa4195fa27600fd43`
Just wanted to make sure we are on the same page, am i getting it right ?

Reply:
🤔 I was also thinking that we could use sprintf to set the salt:

```
$md5$v=1$pf=%[0]s%[1]s$salt$hash
$md5$v=1$hash
```

where `pf` is `printf` and the first argument being the salt, the second the hash. If `pf` is not set, we do not have a salt. This would allow us to define the salt format any way we like:

```
%[0]s%[1]s
%[1]s%[0]s
%[0]sSOME_DELIMITER%[1]s
....
```

WDYT?

Reply:
That's a good idea, this would make salting process more flexible. But i think formatting by index might come as *too implicit* for some users. How about formatting with named parameters like this: 
```
%[salt]s%[password]s
%[salt]sSOME_DELIMITER%[password]s
```
We might replace `%[salt]s` to `%[0]s` and `%[password]s` to `%[1]s` before passing format string to `sprintf`. For example format string `%[salt]s%[password]s` can turn into `%[0]s%[1]s` before passing it to `sprintf`. 
This way we both can use indexed formatting with sprintf and make the definition of format more explicit to the users.

What do you think ?

Reply:
That's true, but I think it makes sense to just rely of `Sprintf`. In v2 we can think about introducing parsing here :)

Reply:
Or we can make `$pf` variable enclosed by double quotes, and only thing the users would have to escape could be the double quotes itself and `%` char. 
For example:
```
<salt>[$<password> => $md5$v=1$pf="%[0]s[$%[1]s"$salt$hash
<salt>"%<password> => $md5$v=1$pf="%[0]s\"%%%[1]s"$salt$hash
```

Reply:
As a third option we may not need any escaping at all. We can assume that part starts after `$pf=` and ends before `$salt` is the format string. For example:
```
<salt>[$<password> => $md5$v=1$pf=%[0]s[$$salt$hash
````
Even though that's an option, users would still need to escape `%` character, because of how Sprintf works.

Reply:
> As a third option we may not need any escaping at all. We can assume that part starts after `$pf=` and ends before `$salt` is the format string. For example:
> 
> ```
> <salt>[$<password> => $md5$v=1$pf=%[0]s[$%[1]s$salt$hash
> ```
> 
> Even though that's an option, users would still have to escape `%` character, because of how Sprintf works.

We can capture the input with this regex: `\$md5\$v\=1\$pf\=(.*)\$(.*)\$(.*)`

Reply:
I'm not sure if we have to escape it! Splitting by `$` and then processing each parameter in it's own context, it should work IMO :) The one problem I see is if the template itself uses `$`: `pf=%[0]s$%[1]s` 🤔 

In which case escaping it in double quotes might make sense: `pf="%[0]s$%[1]s"`

Reply:
I think we are on different page now, so i'm just going to update the code according to thoughts i have in my mind, may be this can illuminate our conversation about the topic.

Reply:
I read the implementation yesterday and I still find the usage of the untrusted format string a little scary. At best it is way more powerful than we need it to be.

How about we just specify the format string as some string that contains `PASS` and `SALT` template strings, and then we `strings.Replace` the actual hash into that? The examples cover all use cases that I can think of (password or salt first, delimiters, etc.), e.g.:

* `PASS:SALT`
* `SALTPASS`
* `PASS$SALT`

To support literal `$`s in the template we can either base64-encode the whole string, or wrap the template with double quotes. I tend towards base64, because this is consistent with other fields in the format (salt and hash), and because the parsing format is then just a `strings.Split(s, "$")`, which is way easier to read than a regular expression.

WDYT?

Reply:
@hperl Definitely +1 on this. This feels way more secure than the `Printf` approach. About the base64 encoded format +1 on this too. This can ease the parsing process. Instead of using regex we can use `Split` approach as you suggested.

Reply:
nice!